app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php

```diff
@@ -74,11 +74,22 @@ public function fetchRates()
      */
     private function convertBatch($data, $currencyFrom, $currenciesTo)
     {
+        $apiKey = $this->scopeConfig->getValue(
+            'currency/currencyconverterapi/api_key',
+            \Magento\Store\Model\ScopeInterface::SCOPE_STORE
+        );
+        if (empty($apiKey)) {
+            $this->_messages[] = __('No API Key was specified or an invalid API Key was specified.');
+            $data[$currencyFrom] = $this->makeEmptyResponse($currenciesTo);
```

Comment:
Returning such response is not a option, please throw some error instead.

Reply:
Let me see how to throw a proper error from there, and change it accordingly.

Reply:
@orlangur by reviewing the code again, I found setting the _messages[] array here seems a good idea. Because the caller will check the _messages[] array and report error to users. For this part I actually cloned the code from the FixerIo class.

Here is a screenshot of the warning message: https://pasteboard.co/IbuUilQ.png

Reply:
Thanks, good to know. Still, we should not produce an empty response.

Reply:
I agree with you. Let me check what kind of response is better for this case later today, and refine the code properly.

Reply:
@orlangur I have changed it to throw a `LocalizedException`, in this way there would be a red error message shown in the UI when the key is empty. [Screenshot](https://pasteboard.co/IbQBulv.png)